### DISCUSSION TITLE
Should freerdp_client_send_wheel_event sent with mouse position?

### DISCUSSION THREAD

**Discussion by jackyzy823** [FreeRDP/FreeRDP#10896] [General]
Should freerdp_client_send_wheel_event sent with mouse  position?

When using [KRDP server](https://invent.kde.org/plasma/krdp) with xfreerdp. I encountered an issue. Mouse wheel scrolling  triggers [screen corners](https://userbase.kde.org/System_Settings/Screen_Edges/en) . After some digging i found it is because the mouse wheel event  ~~is set to position (0,0) of the screen to remote~~  's position is unset . So server applies mouse wheel event on position (0,0) which triggers unwanted behavior.

This is not happened with [KRDC client](https://invent.kde.org/network/krdc).

It could be fixed on server side by remembering previous mouse event poistion. https://invent.kde.org/plasma/krdp/-/merge_requests/73

But i'm not sure this is a client side or server side issue.

Ref:
KRDC send wheel event via freerdp_input_send_mouse_event with position.
https://invent.kde.org/network/krdc/-/blob/master/rdp/rdpsession.cpp#L1076

**Comment by jackyzy823** [comment]
// Might be the first code for wheel event not sending position.
https://github.com/FreeRDP/FreeRDP/commit/dc6d3f022bf5a8aff9e391bcb186f01d05296bc1

**Comment by pnowack** [comment]
Well, if the KDE Project (or rather Arjen Hiemstra) would focus on actually reading the documentation, instead of KopyPasting large code parts from gnome-remote-desktop (which is mostly developed in free-time), then those problems and statements, like "That's rather unexpected.", wouldn't occur.

The FreeRDP-server API side just follows the documentation here: The `xPos` and `yPos` values are only valid in mouse events, when the respective 'move' flag is set. In addition to this, button events require that the button press or release events happen at the `xPos` and `yPos`. In a simplified matter this means, that the `xPos` and `yPos` values are only valid, when no wheel or horizontal wheel event occurs.
The documentation even explicitly mentions, that the `xPos` and `yPos` should be ignored on those two type of events.

> It could be fixed on server side by remembering previous mouse event poistion.

It's not really a fix, but workaround. And this workaround would at least not work for two types of scenarios:
1. The scroll event is the very first event (last mouse position is invalid)
2. KRDP's use case is currently just Remote Assistance. If the mouse is moved on the server side and the helper on the client side just spins the wheel, the mouse just jumps on the old position.

The correct way is to __not__ move the mouse in this situation at all.

**Comment by hardening** [comment]
Well mstsc, the reference client, doesn't send the mouse position with wheel. So server-side must be ready for that...

**Comment by akallabeth** [comment]
well, as mentioned see [\[MS-RDPBCGR\] 2.2.8.1.2.2.3 Fast-Path Mouse Event (TS_FP_POINTER_EVENT)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/16a96ded-b3d3-4468-b993-9c7a51297510) (and the corresponding slowpath equivalent)

**Comment by jackyzy823** [comment]
Thanks for all your replies.

Another question.  I observed that `xfreerdp` sends a PTRFLAGS_MOVE event with current mouse position before each PTRFLAGS_WHEEL event. Is this intentional?

Since i tested that `mstsc` won't do this , it just keeps sending PTRFLAGS_WHEEL event with position (0,0) , and this makes KRDP behaving normally.

-----
Edit
~~This might be KRDP side's problem?  since i test xfreerdp to sfreerdp-server , this doesn't happen.~~ See new comment.

**Reply by akallabeth** [reply]
@jackyzy823 we send whatever `X11` events we get (and usually you do move the mouse a little when scrolling)
so you better be prepared that these events can be mixed (e.g. `move -> scroll -> scroll -> move -> scroll ...`

**Reply by jackyzy823** [reply]
I'm sure the mouse is not moved. I log all the mouse event's position.

I tested  different freerdp client connecting to sfreerdp-server and the result:

Running xfreerdp under wayland (gnome/kde), the situation (sends a PTRFLAGS_MOVE event with current mouse position before each PTRFLAGS_WHEEL event)  will happen.
Running xfreerdp under x11 (cinnamon)  the situationwill not happen.
Running xfreerdp under x11 (gnome)  the situation will happen. (The most weird one)

Running wlfreerdp under wayland (gnome/kde)  the situation will not happen.

**Reply by pnowack** [reply]
It doesn't matter, whether the server side receives a mouse move event with the `PTRFLAGS_MOVE` flag or not before an event with the `PTRFLAGS_WHEEL`- or `PTRFLAGS_HWHEEL`-flag. The server side must deem the position in events with either the `PTRFLAGS_WHEEL`- or `PTRFLAGS_HWHEEL`-flag as an invalid one. Their values (the x- and y-position) can here be any arbitrary value

**Comment by jackyzy823** [comment]
Hi all ,sorry for the dumb question.  I found the root of the problem :( 

There's nothing wrong with FreeRDP side.

KRDP server side has some implementation problem.
